### ISSUE TITLE
pattern match lambda treating Refl as a variable name

### ISSUE THREAD

**Issue by MarcelineVQ** [idris-lang/Idris2#677]
# Steps to Reproduce
Try either DecEq implementation, the latter will have an error.
```idris
data Fo : Type -> Type where
  MkFo : a -> Fo a

DecEq a => DecEq (Fo a) where
  decEq (MkFo x) (MkFo y) with (decEq x y)
    decEq (MkFo x) (MkFo x) | (Yes Refl) = Yes Refl
    decEq (MkFo x) (MkFo y) | (No contra) = No $ \case Refl => contra Refl
                               -- i.e. \r => case r of Refl => contra Refl

DecEq a => DecEq (Fo a) where
  decEq (MkFo x) (MkFo y) with (decEq x y)
    decEq (MkFo x) (MkFo x) | (Yes Refl) = Yes Refl
    decEq (MkFo x) (MkFo y) | (No contra) = No $ \Refl => contra Refl
 -- decEq (MkFo x) (MkFo y) | (No contra) = No $ \Refl => contra ?observeHole
```
# Expected Behavior
That the above two definitions are equivalent for typechecking.
# Observed Behavior
```
Error(s) building file src/Language/Elab/Deriving/DecEq.idr: While processing right hand side of with
block in decEq. When unifying MkFo x = MkFo y and x = y.
Mismatch between: MkFo y and y.
```
Inspecting the context for ?observeHole gives
```idris
 0 a : Type
   x : a
   y : a
   contra : x = y -> Void
   Refl : MkFo x = MkFo y
------------------------------
observeHole : x = y
```
idris seems to be treating Refl as a variable name, which then shadows the later Refl and results in our error where each Refl are considered at the same type.
Writing
```
decEq (MkFo x) (MkFo y) | (No contra) = No $ \Builtin.Refl => contra Refl
```
will solve this and confirms the issue.

Is it possible that the root issue for #633 is also related?

**Comment by gallais** [comment]
I think this is working as expected.

We have lambda case since https://github.com/edwinb/Idris2-boot/pull/244
and other lambdas are not pattern-matching (except for record types?)

Cf. #46 for a similar error

**Comment by Russoul** [comment]
You can pattern match on any constructor if you wrap a pattern in parentheses:

```
testBad : 1 = 1 -> 2 = 2 -- does not compile
testBad = \Refl => Refl -- left Refl is a pattern variable

testWorse : 1 = 1 -> 2 = 2 -- compiles !
testWorse = \(Refl) => Refl -- left Refl is being matched on

```

**Comment by MarcelineVQ** [comment]
@gallais 
lambdacase lets you examine all the cases directly, where `(\case True => ...; False => ...)` and `(\b => case b of True => ...; False => ...)` are equivalent but the pattern matching lambda is slightly different in that `(\True => ...)` becomes `(\x => case x of Foo => ...)` internally, casing still occurs but just over one case. The issue here seems to be that True in that latter lambda form is being treated as a variable instead of a constructor when written without namespace dots or, it turn outs, parens like @Russoul shows.

**Comment by gallais** [comment]
> I think this is working as expected.

I don't agree with my past self now that I have spotted the code in the desugarer
responsible for this behaviour. I do believe it's a bug and it should be fixed by #837.

Thanks for bearing with me. :grin:

**Comment by gallais** [comment]
Closed with 40258965725bf50be632764fe7701244e86cdd74
